common/determined_common/experimental/checkpoint/_torch.py

```diff
@@ -1,25 +1,32 @@
 import pathlib
-from typing import Any, Dict, cast
+from typing import Any, Dict, Union, cast
 
 import torch
 
-from determined.experimental._native import _local_trial_from_context
-from determined.pytorch import PyTorchTrial
+from determined import experimental
+from determined.pytorch import PyTorchTrial, PyTorchTrialContext
 
 
-def load_model(ckpt_dir: pathlib.Path, metadata: Dict[str, Any], **kwargs: Any) -> torch.nn.Module:
-    trial = _local_trial_from_context(
+def load_model(
+    ckpt_dir: pathlib.Path, metadata: Dict[str, Any], **kwargs: Any
+) -> Union[PyTorchTrial, torch.nn.Module]:
```

Comment:
blocking: can we take the following strategy:
* let `load_model` stay the same, but deprecate it.  It will continue to work, but only on old, deprecated PyTorchTrials.
* introduce a `load_trial` which returns a trial.

It seems like a rename needs to happen with this change to the API, because "model" refers to just `nn.Module` objects in our PyTorchAPI, and now we are planning to return more than just `nn.Module` objects.

Reply:
I like this idea. Let me check the possibility of it with @sidneyw

Reply:
It's certainly possible. My worry is that the pytorch checkpoint export would have a different method and behavior from tensorflow in the long run and might be confusing. If we would rather give users a trial class instead of their model we may want to change keras and estimators to do the same.

Reply:
It doesn't sound necessary to make the behavior of different framework checkpoint exporting consistent. When we introduce multiple models to tf.keras, we might need to change the checkpoint interface accordingly. But for now the trial interface is very different. It can be implemented by inheriting `Checkpoint` to be `PytorchCheckpoint`. What do you think?

Reply:
I don't mind adding the `load_trial` call. Why do we need a separate `Checkpoint` class?

Reply:
I mean we only provide `load_model` for keras and estimator not for Pytorch so the classes of different frameworks are different.

Reply:
I don't see the extra subclass as necessary personally. I would just add load_trial to the checkpoint class and support loading a trial from any checkpoint (which we already have the ability to). I would just update the `format` field for flexible pytorch trials. If someone calls load model on a flexible trial they get an error and a helpful message that says use `load_trial`.

Reply:
I'll simply return a trial in `load_model` for now since if this PR doesn't aim at unifying the user experience of different cases. It's fine to keep the current interface and return only the trial for the new PytorchTrial.

Reply:
I'm sorry, but I still feel that returning a `Trial` object from a call named `load_model` is a mistake.  Some day pretty soon we won't support the old-style PyTorchTrial interface anymore, but we'll still have this `load_model` thing around.

In general, most efforts we have made to unify multiple frameworks (the original Trial class with it's `train_batch` callback, the `BatchLoader` interface) have failed because different frameworks behave pretty differently.  Instead, the approach of trying to be idiomatic for each separate framework has proven to be a lot nicer for each framework.

After thinking through it, I actually like having a separate Checkpoint class as @shiyuann suggests, since it seems reasonable to me that different framework checkpoints support different operations.  But I don't feel strongly like we need to implement it that way.

Reply:
`load_model` is actually only used internally and only `Checkpoint.load` is exposed. Even if we remove the old interface, we probably will keep loading from the old schema around.

I think from the users' perspective they want a consistent name for loading a checkpoint as opposed to finding not being able to call `Checkpoint.load` for the checkpoint of the new interface. And IMO in this case, it's fine to unify the name of the loading function. A more well-thought way might be exposing `Checkpoint.load_trial` and `Checkpoint.load_model`, which is breaking and has no obvious benefits for now.

Reply:
Oh, I see.  I did not realize that.  I'm fine with this then.